### DISCUSSION TITLE
Issue while changing log level at runtime

### DISCUSSION THREAD

**Discussion by skeleton18** [gabime/spdlog#2218] [Q&A]
Hello,
I'm trying to change log levels at runtime but it is not working as I expected.
I am using spdlog 1.9.2.

Here is what I have done :

1/ In my header file I have just define `SPDLOG_ACTIVE_LEVEL` to `SPDLOG_LEVEL_TRACE` before including `spdlog/spdlog.h`.

2/ Then I declare 3 loggers:

1. A default console logger `spdlog::stdout_color_mt("default")` set with level `spdlog::level::debug`
2. An error console logger `spdlog::stderr_color_mt("errorLogger")` set with level `spdlog::level::warn`
3. A file logger `spdlog::basic_logger_mt("fileLogger", "logs/basic-log.txt", true)` set with level `spdlog::level::trace`

3/ Finally I wrote a simple main to log messages for each log level (trace, debug, info, warn, error and critical) using `SPDLOG_LOGGER_XXX` macros.

When I run my main program setting `export SPDLOG_LEVEL="*=off,default=warn"` (i.e. disable all loggers but default, and change its log level to warn), it does not seem to work because I had the following output:

> [2021-12-21 11:29:28.664 +01:00] [default] [Logger.h] [debug] [thread #3969739] Testing trace
> [2021-12-21 11:29:28.664 +01:00] [default] [Logger.h] [info] [thread #3969739] Testing trace
> [2021-12-21 11:29:28.664 +01:00] [errorLogger] [Logger.h] [warning] [thread #3969739] Testing trace
> [2021-12-21 11:29:28.665 +01:00] [default] [Logger.h] [warning] [thread #3969739] Testing trace
> [2021-12-21 11:29:28.665 +01:00] [errorLogger] [Logger.h] [error] [thread #3969739] Testing trace
> [2021-12-21 11:29:28.665 +01:00] [default] [Logger.h] [error] [thread #3969739] Testing trace
> [2021-12-21 11:29:28.665 +01:00] [errorLogger] [Logger.h] [critical] [thread #3969739] Testing trace
> [2021-12-21 11:29:28.665 +01:00] [default] [Logger.h] [critical] [thread #3969739] Testing trace

What do I do wrong ?
Thanks for the help.

**Comment by tt4g** [comment]
The first `*=` is not necessary.

```diff
- export SPDLOG_LEVEL="*=off,default=warn"
+ export SPDLOG_LEVEL="off,default=warn"
```

And should call `spdlog::cfg::load_env_levels()` to load log level from environment variable `SPDLOG_LEVEL`.

**Reply by skeleton18** [reply]
Thank you for the quick answer.

Indeed, without the `*=` at the beginning, it works perfectly when I call the `spdlog::cfg::load_env_levels()` function before logging.

Just one more question, let's imagine that I set `SPDLOG_LEVEL` to `"off,default=warn"` and then I launch my service.
If I just want to change at runtime, the log level of `fileLogger`, Can I just set `SPDLOG_LEVEL` to `fileLogger=info` or do I need to set full configuration with `off,default=warn,fileLogger=info`?

**Reply by tt4g** [reply]
>  do I need to set full configuration with `off,default=warn,fileLogger=info`?

Yes. If a global level is specified in `SPDLOG_LEVEL`, then any logger that does not have a logger name and log level pair specified will have its global log level set.

Another thing to note is that `SPDLOG_LEVEL` only accepts a maximum of 512 characters: https://github.com/gabime/spdlog/blob/v1.9.2/include/spdlog/cfg/helpers-inl.h#L82-L87
This restriction makes it difficult to control the log level of many loggers with `SPDLOG_LEVEL`.
